### PR TITLE
fix: octal, hex, and unicode strings at the end of a file

### PR DESCRIPTION
**Pull request by flavorjones**
These were causing invalid memory access.

Also:
- clean up handling of EOF after `"\\"` by moving the check from `parser_lex` into `yp_unescape_calculate_difference`
- report a syntax error for unterminated bracket-style unicode strings, like `\u{1234`

### CODE REVIEW

File: src/unescape.c [hunk 1]
```diff
@@ -236,13 +237,14 @@ unescape(yp_parser_t *parser, uint8_t *dest, size_t *dest_length, const uint8_t
 
                 unicode_cursor += yp_strspn_whitespace(unicode_cursor, end - unicode_cursor);
 
-                while ((*unicode_cursor != '}') && (unicode_cursor < end)) {
+                while ((unicode_cursor < end) && (*unicode_cursor != '}')) {
```
Comment (jemmaissroff) [inline src/unescape.c:240]: Why this change?

**Review by jemmaissroff — COMMENTED** [review]
Looks good, just a small question

File: src/unescape.c (hunk 1 above)
Comment (flavorjones) [reply src/unescape.c:240]: We need to check if `unicode_cursor` is past `end` _before_ dereferencing it, otherwise we could be illegally accessing memory past the end of the buffer.

Comment (flavorjones) [reply src/unescape.c:240]: Hmm, looking at it more closely and this whole loop construct is more complicated than we need. Let me take another pass at cleaning it up.

**Review by jemmaissroff — APPROVED** [review]
LGTM, thanks!
